### PR TITLE
Make crio-lxc working with kubernetes

### PR DESCRIPTION
**Pull request by r10r**
Hi crio-lxc devs,

I've spend some time and made several commit to the crio-lxc source to get it running with kubernetes.
With my changes I can bootstrap a single node kubernetes setup successfuly including the dashboard. I have also deployed several helm chars with success, even those using multiline startup commands (e.g k8s-land/gitea).
( my branch also fixes for #17 - didn't see the pull request before)

Short background:  At my company we are currently running our monolitic system as lxc container and we want
to take advantage of crio/kubernetes features - so we are willing to contribute to crio-lxc in the future.
I'll take some time to add required test cases soon, but I'm currently migrating our application which 
will take some time.

Please take some time and have a look at the changes I made and let's start discussion. 

Greeting,
Ruben

My system setup:

```
root@container-node:~/crio-lxc# uname -a
Linux container-node 5.7.0-1-amd64 #1 SMP Debian 5.7.6-1 (2020-06-24) x86_64 GNU/Linux
root@container-node:~/crio-lxc# cat /etc/debian_version 
bullseye/sid
```

```
kubeadm version: &version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-26T03:45:16Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

```
root@container-node:~/crio-lxc# crio --version
crio version 1.19.0-dev
Version:       1.19.0-dev
GitCommit:     a7b54a80cb5f4818b2175e31fe44a84f6755a5cd
GitTreeState:  clean
BuildDate:     2020-07-16T11:23:53Z
GoVersion:     go1.14.4
Compiler:      gc
Platform:      linux/amd64
Linkmode:      dynamic
```

```
lxc 4.0.2
root@container-node:~/crio-lxc# lxc-checkconfig 
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-5.7.0-1-amd64
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled

--- Control groups ---
Cgroups: enabled

Cgroup v1 mount points: 
/sys/fs/cgroup/systemd
/sys/fs/cgroup/pids
/sys/fs/cgroup/memory
/sys/fs/cgroup/net_cls,net_prio
/sys/fs/cgroup/devices
/sys/fs/cgroup/blkio
/sys/fs/cgroup/cpuset
/sys/fs/cgroup/freezer
/sys/fs/cgroup/cpu,cpuacct
/sys/fs/cgroup/perf_event
/sys/fs/cgroup/rdma

Cgroup v2 mount points: 
/sys/fs/cgroup/unified

Cgroup v1 clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, loaded
CONFIG_NF_NAT_IPV4: missing
CONFIG_NF_NAT_IPV6: missing
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, loaded
FUSE (for use with lxcfs): enabled, loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: 

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/bin/lxc-checkconfig
```

### CODE REVIEW

**Comment by brauner** [comment]
@hallyn Is this something that one of your people can review? :)

**Comment by r10r** [comment]
@brauner I'm currently working on script to bootstrap a kubernetes single node/cluster based on debian buster with crio-lxc  I'll release it soon - this will make a test setup fairly easy for anyone willing to test/review.  

With the latest crio-lxc version the sonobuoy conformance are running against a setup with kubernetes 1.19 have currently 6 failures out of 303 tests.

**Comment by brauner** [comment]
> @brauner I'm currently working on script to bootstrap a kubernetes single node/cluster based on debian buster with crio-lxc I'll release it soon - this will make a test setup fairly easy for anyone willing to test/review.
> 
> With the latest crio-lxc version the sonobuoy conformance are running against a setup with kubernetes 1.19 have currently 6 failures out of 303 tests.

Very impressive. :)

**Comment by hallyn** [comment]
> @hallyn Is this something that one of your people can review? :)

Thanks for this @r10r !

@brauner - yeah,  thanks - @mikemccracken had mentioned having a few questions, but I think he's been a bit swamped.  I made a note to look through a bit myself tomorrow.

File: cmd/create.go [hunk 1]
```diff
@@ -294,9 +294,26 @@ func configureContainer(ctx *cli.Context, c *lxc.Container, spec *specs.Spec) er
...
-		// Make mount paths relative to container root https://github.com/lxc/lxc/issues/2276
-		dest := strings.TrimLeft(ms.Destination, "/")
+
+		// Fix failing mounts due to symlink protection
+		// CVE-2015-1335: Protect container mounts against symlinks
+		// https://github.com/lxc/lxc/commit/592fd47a6245508b79fe6ac819fe6d3b2c1289be
+		// mount targets that contain symlinks must be resolved
+		dest := ms.Destination
+		if strings.HasPrefix(dest, "/var/run") {
+			dest = strings.TrimPrefix(dest, "/var")
```
Comment (hallyn) [inline cmd/create.go]: Hm, how can we do this more generically, without going to the trouble of chrooting into the container rootfs and doing a readlink...   @brauner any suggestions?

**Review by hallyn — CHANGES_REQUESTED** [review]
Note to self - I'm stopping at "Fix mounts on /var/run" .  will pick up from there later.

**Comment by hallyn** [comment]
This would be much easier to review if the commits which are overwritten by later commits were squashed.

If you have time to do that, I'll wait and re-review.

If not, then I'll let @mikemccracken speak up before I merge it.  Mike if you don't have time that's fine, I just don't want to step on your toes and merge if you had pending review comments.

**Comment by r10r** [comment]
@hallyn Sorry for the delay, I'm currently setting up a prototype for our k8s cluster and still have a few fixes/changes for crio-lxc in the pipeline.  I'll try to squash the commits together for a better review.  I'll then force pull the changes to the master branch - is that ok?

File: cmd/create.go (hunk 1 above)
Comment (brauner) [reply cmd/create.go]: What is this trying to do in the first place, i.e. why is this stripping `/var`?

Comment (r10r) [reply cmd/create.go]: This is an bugfix get containers started. /var/run "almost" always symlinks to /run and therefore lxc symlink protection kicks in and refuses the mount. I've implemented proper path resolution later on https://github.com/Drachenfels-GmbH/crio-lxc/blob/d5146786d316e4a1b28d540d3a6b78ffc18d5cb8/cmd/utils.go#L127 
As @hallyn proposed I'll squash my commits to make the review much easier.

**Comment by hallyn** [comment]
> @hallyn Sorry for the delay, I'm currently setting up a prototype for our k8s cluster and still have a few fixes/changes for crio-lxc in the pipeline. I'll try to squash the commits together for a better review. I'll then force pull the changes to the master branch - is that ok?

Please let me know when you've done that :)

**Comment by vrubiolo** [comment]
Just a note to mention that I am interested in those changes as well as I am evaluating using k8s w/ crio-lxc (also to evaluate migrating our application :). Thanks for doing the changes @r10r !

**Comment by r10r** [comment]
Hi @vrubiolo I have a single pending issue with k8s initialization which I want to solve prior to release.
Expect a release within this week (hopefully within the next two days). 
Can you tell me what distro  you use?  I'm currently testing with debian and archlinux.

**Comment by vrubiolo** [comment]
Hi @r10r and thanks for the quick reply ! For the distro, are you asking about which distro is running on the k8s node?

Right now, I don't have a dedicated k8s cluster, I am currently using minikube 1.15.0 on Fedora32. When enabling `cri-o` as the runtime, this uses the Docker driver and an Ubuntu 20.04.1 container into which k8s runs

**Comment by r10r** [comment]
> Hi @r10r and thanks for the quick reply ! For the distro, are you asking about which distro is running on the k8s node?
> 
> Right now, I don't have a dedicated k8s cluster, I am currently using minikube 1.15.0 on Fedora32. When enabling `cri-o` as the runtime, this uses the Docker driver and an Ubuntu 20.04.1 container into which k8s runs

Ok that's good to know. You can help me adapting the deployment script for fedora.

**Comment by vrubiolo** [comment]
> Ok that's good to know. You can help me adapting the deployment script for fedora.

Yes, I'd be glad to !

Btw, if you have an idea about #24 , I'd be most grateful :)

**Comment by r10r** [comment]
> > Ok that's good to know. You can help me adapting the deployment script for fedora.
> 
> Yes, I'd be glad to !
> 
> Btw, if you have an idea about #24 , I'd be most grateful :)

I'll suggest you to wait until tomorrow. There is lot that either doesn't work or is broken in the current version. I'll release tomorrow and you can test again. I'm sure this is already fixed in my version.

**Comment by vrubiolo** [comment]
> I'll suggest you to wait until tomorrow. There is a lot that either doesn't work or is broken in the current version. I'll release tomorrow and you can test again. I'm sure this is already fixed in my version.

Fantastic! Iooking forward to the release then :+1:

**Comment by r10r** [comment]
@vrubiolo You can try to bootstrap your system with the files from https://gist.github.com/r10r/c6c7f43ed85bdcd37fc082b0214c7c6f 
It may contain errors, because I simply copied it from my deployment script. 
Maybe I'll turn this into a separate git repository. 
It will take still an hour or two for the release ;)

**Comment by vrubiolo** [comment]
Hi @r10r , did you get more feedback on your PR? 
I should be able to get back to this in the coming weeks if all goes well (I need to look into building OCI images first)

**Comment by r10r** [comment]
@vrubiolo not yet.  But shame on me - I didn't update the pull request yet.
I did quite a lot of testing and refactoring within the last weeks and I think it was worth it.
Currently I've a single outstanding issue to resolve, which I'm now able to reproduce.
I'm very confident to resolve this before next week.
Just ping me If you don't hear anything.

**Comment by vrubiolo** [comment]
> Just ping me If you don't hear anything.
@r10r : thanks for the update, glad to see things are moving on your end. I'll definitely ping you !

**Comment by r10r** [comment]
Sonobuoy conformance tests are now looping for 2 days successfully on our test clusters.
Please read through the README.md to get an overview of the changes.

I invite everybody to try out our dev branch https://github.com/Drachenfels-GmbH/crio-lxc/tree/v0.9.4

**Comment by r10r** [comment]
@brauner @hallyn  I'm really exited about your feedback. Please try out version v0.9.4. I consider it pre-production ready.

I've done a lot of refactoring lately to improve the structure and design. And I think it was really worth it.
But it lead to a lot of merge commits which are hard to follow. I tried to squash commits more than once - but honestly, it currently takes to much precious time to clean up my commit mess. Commit history is currently still not suited for review. 

But since my desire to refactor is satisfied :D - we should get review started.
My idea is to simply squash everything together and create a single commit for review. What do you think?

@brauner I consider certain parts (cmd/init/crio-lxc-init.c, cmd/start/crio-lxc-start.c, cmd/container-hook/hook.c) for inclusion in lxc itself. Please let me know what you think about it.

**Comment by r10r** [comment]
@vrubiolo Please report any issues to https://github.com/Drachenfels-GmbH/crio-lxc until merge is complete. Thanks!

**Comment by vrubiolo** [comment]
Hi @r10r , just a note that I am back on this and now looking/using your fork at https://github.com/Drachenfels-GmbH/crio-lxc instead.
Thanks for all the changes you put in to allow k8s, it looks like the project was quite far from enabling it fully given all the commits you added...

**Comment by r10r** [comment]
I'll close this pull request now, since my fork is now a complete rewrite and shares very little code with crio-lxc.
I renamed the project to [**lxcri**](https://github.com/drachenfels-de/lxcri) for a clear distinction between the projects.

The project has still some rough edges and had to endure my refactorings.
With liblxc from master sonobuoy conformance tests are now passing 100%. 
A public API is in progress, but is currently missing documentation and examples.
I'll release docker images soon - so anyone who wants to try out a kubernetes cluster
backed by liblxc should have look at it.

**Pull request closed without merging**